BosunReporter/Handlers/DataDogStatsdMetricHandler.cs

```diff
@@ -0,0 +1,294 @@
+﻿using BosunReporter.Infrastructure;
+using Pipelines.Sockets.Unofficial;
+using Pipelines.Sockets.Unofficial.Buffers;
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BosunReporter.Handlers
+{
+    /// <summary>
+    /// Implements <see cref="BufferedMetricHandler" /> by sending data to a DataDog agent.
+    /// </summary>
+    public class DataDogStatsdMetricHandler : BufferedMetricHandler
+    {
+        readonly Lazy<ValueTask<IPEndPoint>> _endpointFactory;
+        readonly Socket _socket;
+        readonly Lazy<BufferWriter<byte>> _metricBufferWriter;
+        readonly Lazy<BufferWriter<byte>> _metadataBufferWriter;
+
+        const int ValueDecimals = 5;
+        static readonly byte[] s_counter = Encoding.UTF8.GetBytes("c");
+        static readonly byte[] s_gauge = Encoding.UTF8.GetBytes("g");
+        static readonly byte[] s_pipe = Encoding.UTF8.GetBytes("|");
+        static readonly byte[] s_colon = Encoding.UTF8.GetBytes(":");
+        static readonly byte[] s_newLine = Encoding.UTF8.GetBytes("\n");
+        static readonly byte[] s_hash = Encoding.UTF8.GetBytes("#");
+        static readonly byte[] s_comma = Encoding.UTF8.GetBytes(",");
+        static readonly StandardFormat s_valueFormat = StandardFormat.Parse("F" + ValueDecimals);
+
+        /// <summary>
+        /// Constructs a new <see cref="DataDogStatsdMetricHandler" /> pointing at the specified <see cref="Uri" />.
+        /// </summary>
+        /// <param name="host">
+        /// Host of a DataDog agent.
+        /// </param>
+        /// <param name="port">
+        /// Port of a DataDog agent.
+        /// </param>
+        public DataDogStatsdMetricHandler(string host, ushort port)
+        {
+            if (host == null)
+            {
+                throw new ArgumentNullException(nameof(host));
+            }
+
+            _endpointFactory = new Lazy<ValueTask<IPEndPoint>>(() => CreateEndpointAsync(host, port));
+            _socket = new Socket(SocketType.Dgram, ProtocolType.Udp);
+
+            BufferWriter<byte> CreateBufferWriter() => BufferWriter<byte>.Create(MaxPayloadSize);
+
+            _metricBufferWriter = new Lazy<BufferWriter<byte>>(CreateBufferWriter);
```

Comment:
allocates two delegates... frankly I'd lose the `Lazy<T>` here, but if you want to use it, do something like: `private static readonly Func<BufferWriter<byte>> s_CreateBufferWriter = () => BufferWriter<byte>.Create(MaxPayloadSize);` and pass `s_CreateBufferWriter` to lazy.

Reply:
hmmm... maybe this isn't possible because of the `this`, in which case... frankly I'd be back at the "lose the `Lazy<T>`"